### PR TITLE
Support Dockerfile.dockerignore

### LINKED ISSUE #776: Support Dockerfile.dockerignore
Currently kaniko only looks for a `.dockerignore` file but a uselful feature of `docker build` is to take `<path/to/Dockerfile>.dockerignore` into account if it exists.

See https://github.com/moby/buildkit/pull/901

### PR DESCRIPTION
**Pull request by victornoel**
This fixes #776

**Description**

- Use `Dockerfile.dockerignore` in priority over the context's `.dockerignore` file.
- Actual name checked is based on the docker file passed to the command line
- Introduces unit tests and integration test to validate the behaviour.

~~I haven't added an unit test because it wasn't clear what was to correct way to test this because excludes seems to be managed globally via a static field. I would be interested by guidance from knowledgeable people on this.~~

**Submitter Checklist**

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [unit tests](../DEVELOPMENT.md#creating-a-pr)
- [x] Adds integration tests if needed.

_See [the contribution guide](../CONTRIBUTING.md) for more details._

**Reviewer Notes**

- [x] The code flow looks good. 
- [x] Unit tests and or integration tests added.

**Release Notes**

- Support `Dockerfile.dockerignore` file in the same way as `docker build`

### CODE REVIEW

File: pkg/util/fs_util_test.go [hunk 1]
```diff
@@ -694,3 +694,49 @@ func Test_childDirInWhitelist(t *testing.T) {
...
+		name string
+		args args
+	}{
+		{
+			name: "relative dockerfile used",
+			args: args{
+				dockerfilepath: "../../integration/dockerfiles/Dockerfile_test_dockerignore_relative",
+				buildcontext:   "../../integration/",
+				excluded:       "ignore_relative/bar",
+				notExcluded:    "ignore_relative/foo",
```
Comment (tejal29) [inline pkg/util/fs_util_test.go]: i think we should test `ignore/bar` is not excluded along with `ignore_relative/foo` here to make sure context dockerignore is not used.? What do you think?

File: pkg/util/fs_util_test.go [hunk 2]
```diff
@@ -694,3 +694,49 @@ func Test_childDirInWhitelist(t *testing.T) {
...
+			},
+		},
+		{
+			name: "context dockerfile is used",
+			args: args{
+				dockerfilepath: "../../integration/dockerfiles/Dockerfile_test_dockerignore",
+				buildcontext:   "../../integration/",
+				excluded:       "ignore/bar",
+				notExcluded:    "ignore/foo",
```
Comment (tejal29) [inline pkg/util/fs_util_test.go]: We should test `ignore_relative/bar` is not excluded here to make sure `Dockerfile_test_dockerignore_relative.dockerignore` is not used?

**Comment by tejal29** [comment]
The newly added dockerfile failed
```
Error building images: Failed to build image gcr.io/kaniko-test/docker-dockerfile_test_dockerignore_relative with docker command "[docker build -t gcr.io/kaniko-test/docker-dockerfile_test_dockerignore_relative -f dockerfiles/Dockerfile_test_dockerignore_relative .]": exit status 1 Sending build context to Docker daemon  122.4kB

Step 1/2 : FROM scratch
 ---> 
Step 2/2 : COPY ignore_relative/* /foo
When using COPY with more than one source file, the destination must be a directory and end with a /
exit status 1
FAIL	github.com/GoogleContainerTools/kaniko/integration	265.988s

```

**Comment by victornoel** [comment]
@tejal29 apparently, docker only reads the `Dockerfile.dockerignore` when `DOCKER_BUILDKIT=1` is used for `docker build`: https://github.com/moby/moby/issues/12886#issuecomment-523706042

I'm not sure what is the best course of action then now. Is there some way to alter the integration test for this one or should we just drop it because it's not supported by vanilla docker build?

File: pkg/util/fs_util_test.go (hunk 1 above)
Comment (victornoel) [reply pkg/util/fs_util_test.go]: very good point, yes!

**Comment by victornoel** [comment]
@tejal29 I have another more unit tests (and refactored it to separate them in multiple `t.Run`) per your comments

File: integration/dockerfiles/Dockerfile_test_dockerignore_relative [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# This dockerfile makes sure Dockerfile.dockerignore is working
```
Comment (tejal29) [inline integration/dockerfiles/Dockerfile_test_dockerignore_relative]: Can we remove this from integration test and rename this to "Dockerfile_dockeringore_relative"

**Comment by tejal29** [comment]
Thanks @victornoel, your integration test is still going to fail. 


> @tejal29 I added another commit with more unit tests (and refactored it to separate them in multiple `t.Run`) per your comments

Awesome!

The integration test will be fixed if you rename `Dockerfile_test_dockerignore_relative` to not match pattern https://github.com/GoogleContainerTools/kaniko/blob/22188595f01c452c376db7561368b5b49272ad37/integration/images.go#L99

File: integration/dockerfiles/Dockerfile_test_dockerignore_relative (hunk 3 above)
Comment (victornoel) [reply integration/dockerfiles/Dockerfile_test_dockerignore_relative]: @tejal29 I'm not 100% sure how things work: if the file doesn't have the word `test` in it, then it won't be included in integration tests or is there more to do?

Comment (victornoel) [reply integration/dockerfiles/Dockerfile_test_dockerignore_relative]: ah you answered below, thx :)

**Comment by victornoel** [comment]
@tejal29 it's pushed, do you want me to squash everything into one commit or should I leave it like that?

**Comment by victornoel** [comment]
@tejal29 @kokoro-team seams to have remove the `kokoro:run` tag just after you added it :)

**Comment by tejal29** [comment]
@victornoel, whenever kokoro (internal jenkins) receives the notification via label, it kicks off a job to run the tests on the PR and removes the label.

**Comment by victornoel** [comment]
@tejal29 ah, ok, excellent, thank you for your help, my first PR here and in Go :)

**Comment by guillaume86** [comment]
Could this be upstreamed to moby/moby ? 
I would love to have this option available for my windows builds.
As noted here it would probably get approved: https://github.com/moby/moby/issues/12886#issuecomment-582660722

**Comment by victornoel** [comment]
@guillaume86 isn't what https://github.com/moby/buildkit/pull/901 was for? (cited in https://github.com/GoogleContainerTools/skaffold/pull/3837)

**Comment by guillaume86** [comment]
@victornoel Yes but buildkit is not supported on windows.

**Comment by victornoel** [comment]
@guillaume86 ah, sorry about that, well here it's about kaniko, you would need to ask in the moby repository for implementing this feature, not sure you will get help here :)

**Pull request merged**